File: MergeSort.py
class MergeSort:
    def __init__(self, array):
        self.array = array

    def merge_sort(self):
        if len(self.array) > 1:
            mid = len(self.array) // 2
            left_half = self.array[:mid]
            right_half = self.array[mid:]

            left_merge_sort = MergeSort(left_half)
            right_merge_sort = MergeSort(right_half)

            left_merge_sort.merge_sort()
            right_merge_sort.merge_sort()

            i = j = k = 0

            while i < len(left_half) and j < len(right_half):
                if left_half[i] < right_half[j]:
                    self.array[k] = left_half[i]
                    i += 1
                else:
                    self.array[k] = right_half[j]
                    j += 1
                k += 1

            while i < len(left_half):
                self.array[k] = left_half[i]
                i += 1
                k += 1

            while j < len(right_half):
                self.array[k] = right_half[j]
                j += 1
                k += 1

    def get_sorted_array(self):
        return self.array


# Example usage:
if __name__ == "__main__":
    unsorted_array = [12, 11, 13, 5, 6, 7]
    merge_sort_instance = MergeSort(unsorted_array)
    merge_sort_instance.merge_sort()
    sorted_array = merge_sort_instance.get_sorted_array()

    print("Unsorted array:", unsorted_array)
    print("Sorted array:", sorted_array)

File: MergeSort.pl
% Merge Sort implementation in Prolog

% Merge two sorted lists into a single sorted list
merge([], List, List).
merge(List, [], List).
merge([X|Xs], [Y|Ys], [X|Result]) :-
    X =< Y,
    merge(Xs, [Y|Ys], Result).
merge([X|Xs], [Y|Ys], [Y|Result]) :-
    X > Y,
    merge([X|Xs], Ys, Result).

% Split a list into two halves
split(List, Left, Right) :-
    length(List, Len),
    Half is Len // 2,
    length(Left, Half),
    append(Left, Right, List).

% Merge Sort
merge_sort([], []).
merge_sort([X], [X]).
merge_sort(Unsorted, Sorted) :-
    Unsorted = [_,_|_], % Make sure the list has at least two elements
    split(Unsorted, Left, Right),
    merge_sort(Left, LeftSorted),
    merge_sort(Right, RightSorted),
    merge(LeftSorted, RightSorted, Sorted).

% Example usage:
% ?- merge_sort([12, 11, 13, 5, 6, 7], Sorted).
% Sorted = [5, 6, 7, 11, 12, 13].
